### PR TITLE
Implement SdmxCatalogGroup + SdmxCatalogItem

### LINKED ISSUE #3524: Model Architecture: Implement SdmxCatalogItem

### PR DESCRIPTION
**Pull request by nf-s**
### Implement SdmxCatalogGroup + SdmxCatalogItem

Fixes https://github.com/TerriaJS/terriajs/issues/3524

See [Issue](https://github.com/TerriaJS/terriajs/issues/3524) for progress notes

**To test ->** http://ci.terria.io/sdmx-next/#clean&https://gist.githubusercontent.com/nf-s/24b9288c6cdb6ba41d12f3d269318a0d/raw/c5ed64071c8a651ae458b4c0d3dc79c8618cb86d/sdmx.json

Example catalog:

https://gist.githubusercontent.com/nf-s/24b9288c6cdb6ba41d12f3d269318a0d/raw/c5ed64071c8a651ae458b4c0d3dc79c8618cb86d/sdmx.json

```json
{
  "homeCamera": {
    "north": -8,
    "east": 158,
    "south": -45,
    "west": 109
  },
  "catalog": [
    {
      "type": "sdmx-group",
      "name": "SPC",
      "url": "https://stats-nsi-stable.pacificdata.org/rest",
      "conceptOverrides": [
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=SPC:CS_COMMON(2.0).GEO_PICT",
          "type": "region",
          "regionType": "CNT2"
        }
      ]
    },
    {
      "type": "sdmx-group",
      "name": "ABS",
      "url": "https://api.data.abs.gov.au",
      "conceptOverrides": [
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_C16_COMMON(1.0.0).REGION_TYPE",
          "type": "region-type",
          "selectedId": "SA3"
        },
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_C16_COMMON(1.0.0).REGION",
          "type": "region"
        },
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_C16_COMMON(1.0.0).REGION_SA1",
          "type": "region",
          "regionType": "SA1"
        },
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_GEOGRAPHY(1.0.0).STATE_TERR",
          "type": "region",
          "regionType": "STE_2016"
        },
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_GEOGRAPHY(1.0.0).IND_REGION",
          "type": "region",
          "regionType": "IREG"
        },
        {
          "id": "urn:sdmx:org.sdmx.infomodel.conceptscheme.Concept=ABS:CS_C16_COMMON(1.0.0).STATE",
          "disable": true,
          "allowUndefined": true
        }
      ]
    }
  ]
}
```

### Changes
* Add SDMX-json support for groups and items (using SDMX-csv for data queries)
* `TableMixin` now uses `ExportableMixin` and `AsyncMappableMixin`
* Move region provider loading in `TableMixin` `forceLoadTableMixin` to `loadRegionProviderList`
* Added `TableAutomaticStylesStratum.stratumName` instead of hard-coded strings
* Added `Dimension` interface for `SelectableDimension` - which can be used for Traits
* Make `SelectableDimension.options` optional

### Checklist

-   [x] Simplify generated type definitions
-   [x] Support nested categories
-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Comment by rowanwins** [comment]
There seems to be some duplicate requesting for data going on ATM
Screen_2020_09_21

Also we're re-requesting stuff when the item is opened and closed in the catalog.

**Comment by rowanwins** [comment]
Also getting FeatureInfo doesn't seem to be working for the SPC stuff.

**Comment by nf-s** [comment]
I have sorted duplicate requesting, but not feature info

**Comment by rowanwins** [comment]
I think we can proceed without the feature info stuff as it's likely not directly related to this PR so we can tackle it separately.

**Comment by rowanwins** [comment]
You've managed to cull one of the requests, so it's now only a duplicate rather than a triplicate!
Screen_2020_09_22
From what I can tell it only still occurs for those datasets that don't have regions...

**Comment by rowanwins** [comment]
And opening and closing the catalog group still is making duplicate requests. I encountered similar stuff with my `CkanCatalogGroup` so I added set out my logic like so https://github.com/TerriaJS/terriajs/blob/next/lib/Models/CkanCatalogGroup.ts#L265-L291

**Comment by nf-s** [comment]
@rowanwins fixed the data loading this time (I think)!
Something to do with how `TableMixin` `forceLoadMapItems` and `forceLoadChartItems` is called twice

**Comment by nf-s** [comment]
> 
> 
> I think we can proceed without the feature info stuff as it's likely not directly related to this PR so we can tackle it separately.

And yep, I think it is to do with region mapping

File: lib/Models/SdmxJson/SdmxJsonDataflowStratum.ts [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        catalogItem,
+        `${catalogItem.url}/dataflow/${catalogItem.agencyId}/${catalogItem.dataflowId}?references=all`
+      ),
+      false
+    );
+
+    // Check response
+    if (!isDefined(dataflowStructure.data)) {
+      throw new TerriaError({
+        title: `Could not load SDMX Dataflow`,
```
Comment (rowanwins) [inline lib/Models/SdmxJson/SdmxJsonDataflowStratum.ts]: I know it's painful but do we want to internationalize these strings?

Comment (nf-s) [reply lib/Models/SdmxJson/SdmxJsonDataflowStratum.ts]: yep, no problem

**Review by rowanwins — APPROVED** [review]
Looks really good.

**Pull request merged**